designer/test/component-create.test.js

```diff
@@ -28,17 +35,20 @@ suite('Component create', () => {
     expect(wrapper.find('ComponentTypeEdit').exists()).to.equal(false)
   })
 
-  test('Selecting a component type should display the ComponentTypeEdit component', () => {
-    const wrapper = shallow(<ComponentCreate data={data} page={page} />)
-    const form = wrapper.find('form')
-
-    form.find('select').simulate('change', { target: { value: 'TextField' } })
-
-    const componentTypeEdit = wrapper.find('ComponentTypeEdit')
-    expect(componentTypeEdit.exists()).to.equal(true)
-    expect(componentTypeEdit.prop('page')).to.equal(page)
-    expect(componentTypeEdit.prop('component')).to.equal({ type: 'TextField' })
-    expect(componentTypeEdit.prop('data')).to.equal(data)
-    expect(Object.keys(componentTypeEdit.props()).length).to.equal(3)
+  test('Selecting a component type should display the ComponentTypeEdit component', async flags => {
+    const onInit = () => {
+      const form = wrapper.find('form')
+
+      form.find('select').simulate('change', { target: { value: 'TextField' } })
+
+      const componentTypeEdit = wrapper.find('ComponentTypeEdit')
+      expect(componentTypeEdit.exists()).to.equal(true)
+      expect(componentTypeEdit.prop('page')).to.equal(page)
+      expect(componentTypeEdit.prop('component')).to.equal({ type: 'TextField', name: generatedId })
+      expect(componentTypeEdit.prop('data')).to.equal(data)
+      expect(Object.keys(componentTypeEdit.props()).length).to.equal(3)
+    }
+    const wrappedOnInit = flags.mustCall(onInit, 1)
+    const wrapper = shallow(<ComponentCreate data={data} page={page} onInit={wrappedOnInit} />)
```

Comment:
can I asked why you're passing in expects to a component?

Reply:
It's the only way to check that the onInit actually gets called - i.e. the only way I can find to test async behaviour in hapi lab. I can't assert the layout properly before the id is retrieved and the wrappedOnInit allows time for that to happen

Reply:
possibly this? https://sinonjs.org/how-to/lolex-async-promises/ 

I don't think you should be hijacking `onInit` to make a test work (or is this the purpose of onInit? -- if so, it shouldn't) you should set up a spy for `onInit` and run the test in the test (not in the component)

Reply:
I tried this and can't see how it helps. I've made the componentDidMount function async instead and awaited it in the test to ensure it's been executed.